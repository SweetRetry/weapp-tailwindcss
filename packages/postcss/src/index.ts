import type { IStyleHandlerOptions } from './types'
import { defuOverrideArray } from '@weapp-tailwindcss/shared'
import postcss from 'postcss'
import { getPlugins } from './plugins'

export { createInjectPreflight } from './preflight'

export { internalCssSelectorReplacer } from './shared'

export function styleHandler(rawSource: string, options: IStyleHandlerOptions) {
  return postcss(getPlugins(options))
    .process(
      rawSource,
      options.postcssOptions?.options ?? {
        from: undefined,
      },
    )
    .async()
}

export function createStyleHandler(options: Partial<IStyleHandlerOptions>) {
  return (rawSource: string, opt?: Partial<IStyleHandlerOptions>) => {
    return styleHandler(
      rawSource,
      defuOverrideArray<IStyleHandlerOptions, Partial<IStyleHandlerOptions>[]>(opt as IStyleHandlerOptions, options),
    )
  }
}
