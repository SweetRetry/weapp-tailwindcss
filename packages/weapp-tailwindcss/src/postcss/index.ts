export * from '@weapp-tailwindcss/postcss'
