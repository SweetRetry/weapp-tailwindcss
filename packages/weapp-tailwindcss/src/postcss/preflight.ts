import type { CssPreflightOptions } from '../types'
import type { IPropValue } from '../types/postcss'

export type InjectPreflight = () => IPropValue[]

export function createInjectPreflight(options?: CssPreflightOptions): InjectPreflight {
  const result: IPropValue[] = []
  // if options false ,do no thing
  if (options && typeof options === 'object') {
    const entries = Object.entries(options)
    for (const [prop, value] of entries) {
      if (value !== false) {
        result.push({
          prop,
          value: value.toString(),
        })
      }
    }
  }

  return () => {
    return result
  }
}
