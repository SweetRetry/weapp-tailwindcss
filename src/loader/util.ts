import fs from 'fs'
import path from 'path'

export function mkfileSync (filename: string, content: string) {
  try {
    const dirname = path.dirname(filename)
    if (!fs.existsSync(dirname)) {
      fs.mkdirSync(dirname, {
        recursive: true
      })
    }
    fs.writeFileSync(filename, content, 'utf-8')
  } catch (error) {
    console.error(error)
  }
}
