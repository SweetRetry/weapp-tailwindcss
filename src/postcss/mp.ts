import type { Rule, AtRule } from 'postcss'
import { cssSelectorReplacer } from './shared'

import type { InjectPreflight } from './preflight'
function isSupportedRule (selector: string) {
  return !selector.includes(':hover')
}
const PATTERNS = [/:not\(template\)\s*~\s*:not\(template\)/.source, /:not\(\[hidden\]\)\s*~\s*:not\(\[hidden\]\)/.source].join('|')
const BROAD_MATCH_GLOBAL_REGEXP = new RegExp(PATTERNS, 'g')
// ':not(template) ~ :not(template)'
// ':not(template)~:not(template)'
// const regexp1 = /:not\(template\)\s*~\s*:not\(template\)/g
// :not([hidden])~:not([hidden])
// :not([hidden]) ~ :not([hidden])
// const regexp2 = /:not\(\[hidden\]\)\s*~\s*:not\(\[hidden\]\)/g

export function commonChunkPreflight (node: Rule, cssInjectPreflight: InjectPreflight) {
  node.selector = node.selector.replace(BROAD_MATCH_GLOBAL_REGEXP, 'view + view').replace(/\*/g, 'view')

  // 变量注入和 preflight
  if (/::before/.test(node.selector) && /::after/.test(node.selector)) {
    const selectorParts = node.selector.split(',')
    if (!selectorParts.includes('view')) {
      selectorParts.push('view')
      node.selector = selectorParts.join(',')
    }

    // node.walkDecls((decl) => {
    //   // remove empty var 来避免压缩css报错
    //   if (/^\s*$/.test(decl.value)) {
    //     decl.remove()
    //   }
    //   //  console.log(decl.prop, decl.value)
    // })

    // preset
    if (typeof cssInjectPreflight === 'function') {
      node.append(...cssInjectPreflight())
    }
  }
}

export function mpRulePreflight (node: Rule) {
  // console.log(node.selector)
  if (!isSupportedRule(node.selector)) {
    node.remove()
    return
  }

  node.selector = cssSelectorReplacer(node.selector)

  // node.walkDecls((decl) => {
  //   if (decl.prop === 'visibility') {
  //     switch (decl.value) {
  //       case 'hidden':
  //         decl.replaceWith(decl.clone({ value: 'collapse' }))
  //         return
  //     }
  //   }

  //   if (decl.prop === 'vertical-align') {
  //     switch (decl.value) {
  //       case 'middle':
  //         decl.replaceWith(decl.clone({ value: 'center' }))
  //     }
  //   }
  // })
}

export function mpAtRulePreflight (node: AtRule) {
  // if (node.name === 'media') {
  // }
  // do nothing
}
