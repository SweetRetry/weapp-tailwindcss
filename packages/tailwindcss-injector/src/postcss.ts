import type { PluginCreator } from 'postcss'
import type { Config } from 'tailwindcss'
import { defu } from 'defu'
import postcss from 'postcss'
// import set from 'set-value'
import tailwindcss from 'tailwindcss'

// function isObject(obj: any): obj is object {
//   return typeof obj === 'object' && obj !== null
// }

// function isTailwindcssConfig(obj: any): obj is Config {
//   return isObject(obj)
// }

type InlineTailwindcssOptions = string | Config | {
  config: string | Config
} | undefined

export interface Options {
  filter: (input?: postcss.Input) => boolean
  config: InlineTailwindcssOptions | ((input?: postcss.Input) => InlineTailwindcssOptions)
  directiveParams: ('base' | 'components' | 'utilities' | 'variants')[]
}

const creator: PluginCreator<Partial<Options>> = (options) => {
  const { config, filter, directiveParams } = defu<Options, Options[]>(options, {
    filter: () => true,
    config: undefined,
    directiveParams: ['utilities'],
  })
  return {
    postcssPlugin: 'postcss-tailwindcss-injector',
    async Once(root, helpers) {
      if (filter(root.source?.input)) {
        let _root = root
        for (const params of directiveParams) {
          const node = root.nodes.find(x => x.type === 'atrule' && x.params === params)
          if (!node) {
            _root = _root.prepend(helpers.atRule({ name: 'tailwind', params }))
          }
        }

        const cfg = typeof config === 'function' ? config(root.source?.input) : config
        // if (isTailwindcssConfig(cfg)) {
        //   if ('config' in cfg) {

        //   }
        //   else {
        //     cfg.content = []
        //   }
        // }
        const { root: newRoot } = await postcss([
          tailwindcss(cfg),
        ]).process(_root)

        root.replaceWith(newRoot)
      }
    },
  }
}
creator.postcss = true

export default creator
