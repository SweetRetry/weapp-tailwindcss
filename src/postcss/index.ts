import postcss from 'postcss'

import { mpRulePreflight, commonChunkPreflight } from './mp'
// mpAtRulePreflight
const isMp = true
export function styleHandler (rawSource: string, file: string) {
  const root = postcss.parse(rawSource)
  // console.log(file)
  root.walk((node, idx) => {
    if (node.type === 'rule') {
      if (isMp) {
        // 引用传递
        // uni-app common-> main.wxss
        // taro app.wxss
        if (file === 'main' || file === 'app') {
          commonChunkPreflight(node)
        }
        mpRulePreflight(node)
      }
    }
    //  else if (node.type === 'atrule') {
    //   if (isMp) {
    //     mpAtRulePreflight(node)
    //   }
    // }
    // 顺其自然
    // else if (node.type === 'comment') {
    //   node.remove()
    // }
  })
  return root.toString()
}
