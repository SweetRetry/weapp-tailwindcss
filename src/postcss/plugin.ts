import type { PluginCreator } from 'postcss'
import type { InternalPostcssOptions, StyleHandlerOptions } from '@/types'
import { transformSync } from './selectorParser'
// import { getOptions } from '../defaults'
import { getOptions } from '@/defaults'
import { commonChunkPreflight } from './mp'
import { createInjectPreflight } from './preflight'
const postcssPlugin = 'postcss-weapp-tailwindcss-rename'

export type PostcssWeappTailwindcssRename = PluginCreator<InternalPostcssOptions>

const plugin: PostcssWeappTailwindcssRename = (options: InternalPostcssOptions = {}) => {
  const mergedOptions = getOptions(options)
  const { cssPreflight, cssPreflightRange, customRuleCallback, replaceUniversalSelectorWith } = mergedOptions
  const cssInjectPreflight = createInjectPreflight(cssPreflight)
  const opts: StyleHandlerOptions = {
    cssInjectPreflight,
    cssPreflightRange,
    isMainChunk: true,
    customRuleCallback,
    replaceUniversalSelectorWith
  }

  // eslint-disable-next-line no-unused-vars
  // const { mainCssChunkMatcher } = getOptions<InternalPostcssOptions>(options)
  return {
    postcssPlugin,
    Once (css) {
      // const source = css.source
      // const filePath = source!.input.file as string
      // // if (mainCssChunkMatcher(filePath)) {
      //   css.walkRules((rule) => {
      //     commonChunkPreflight(rule)
      //   })
      // }
      css.walkRules((rule) => {
        transformSync(rule, opts)
        // mpRulePreflight(rule, opts)
        commonChunkPreflight(rule, opts)
      })
    }
  }
}

plugin.postcss = true

export default plugin // as Plugin
