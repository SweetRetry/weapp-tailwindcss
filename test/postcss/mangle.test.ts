import { styleHandler } from '@/postcss'
import ClassGenerator from '@/mangle/classGenerator'
import { cssCasePath, createGetCase, createPutCase } from '../util'
// import { getOptions } from '@/defaults'
const getCase = createGetCase(cssCasePath)
// @ts-ignore
// eslint-disable-next-line no-unused-vars
const putCase = createPutCase(cssCasePath)
describe('[postcss] mangle', () => {
  it('mangle case with empty classGenerator', async () => {
    // const options = getOptions()
    const testCase = await getCase('mangle-nested-0.css')
    const classGenerator = new ClassGenerator()

    const result = styleHandler(testCase, {
      cssInjectPreflight: () => [],
      cssPreflightRange: 'view',
      isMainChunk: true,
      replaceUniversalSelectorWith: 'view',
      classGenerator
    })
    expect(result).toMatchSnapshot()
  })

  it('mangle case 0', async () => {
    // const options = getOptions()
    const testCase = await getCase('mangle-nested-0.css')
    const classGenerator = new ClassGenerator()
    const rawMatrix = [
      ['el-tree-node__content', 'a'],
      ['el-tree-node__expand-icon', 'b']
    ]
    const newClassMap = rawMatrix.reduce<typeof classGenerator.newClassMap>((acc, [key, name]) => {
      acc[key] = {
        name,
        usedBy: []
      }
      return acc
    }, {})
    classGenerator.newClassMap = newClassMap
    classGenerator.newClassSize = rawMatrix.length
    const result = styleHandler(testCase, {
      cssInjectPreflight: () => [],
      cssPreflightRange: 'view',
      isMainChunk: true,
      replaceUniversalSelectorWith: 'view',
      classGenerator
    })
    expect(result).toMatchSnapshot()
  })

  it('mangle-nested-with-comment', async () => {
    // const options = getOptions()
    const testCase = await getCase('mangle-nested-with-comment.css')
    const classGenerator = new ClassGenerator()
    const rawMatrix = [
      ['el-tree-node__content', 'a'],
      ['el-tree-node__expand-icon', 'b']
    ]
    const newClassMap = rawMatrix.reduce<typeof classGenerator.newClassMap>((acc, [key, name]) => {
      acc[key] = {
        name,
        usedBy: []
      }
      return acc
    }, {})
    classGenerator.newClassMap = newClassMap
    classGenerator.newClassSize = rawMatrix.length
    const result = styleHandler(testCase, {
      cssInjectPreflight: () => [],
      cssPreflightRange: 'view',
      isMainChunk: true,
      replaceUniversalSelectorWith: 'view',
      classGenerator
    })
    expect(result).toMatchSnapshot()
  })
})
