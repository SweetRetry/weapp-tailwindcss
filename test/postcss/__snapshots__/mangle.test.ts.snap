// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[postcss] mangle mangle case 0 1`] = `
".a {
  display: flex;
  align-items: center;
}

.a>.b {
  padding: 6px;
}

.a .b::after {
  content: "after";
}

.a label {
  margin-right: 8px;
}
/* mangle disabled */
.el-tree-node__content {
  color: red;
}
.happy {
  word: "have a nice day";
}
"
`;

exports[`[postcss] mangle mangle case with empty classGenerator 1`] = `
".el-tree-node__content {
  display: flex;
  align-items: center;
}

.el-tree-node__content>.el-tree-node__expand-icon {
  padding: 6px;
}

.el-tree-node__content .el-tree-node__expand-icon::after {
  content: "after";
}

.el-tree-node__content label {
  margin-right: 8px;
}
/* mangle disabled */
.el-tree-node__content {
  color: red;
}
.happy {
  word: "have a nice day";
}
"
`;

exports[`[postcss] mangle mangle-nested-with-comment 1`] = `
"/* mangle disabled 222 */
.el-tree-node__content {
  color: red;
}

/* mangle ignore 111 */
.el-tree-node__content {
  color: yellow;
}

.a {
  color: blue;
}
"
`;
